### PR TITLE
Add periodic events with peer and media statistics

### PR DESCRIPTION
**Pull request by davibe**

### CODE REVIEW

File: str0m/src/stats.rs [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+    pub tx: Bytes,
+    pub rx: Bytes,
+    pub ts: Instant,
+}
+
+impl StatsSnapshot {
+    fn new(ts: Instant) -> StatsSnapshot {
+        // TODO: is there a way to only provide the `ts` property
+        // which does not conform to Default,
+        // and have the rest generated by ..Default::default() ?
```
Comment (k0nserv) [inline str0m/src/stats.rs:29]: Unfortunately not, I think I recall seeing an RFC to address this but I can't find it

Comment (davibe) [reply str0m/src/stats.rs:31]: amazing, thx

File: str0m/Cargo.toml [hunk 2]
```diff
@@ -9,6 +9,8 @@ tracing = "0.1"
 rand = "0.8"
 bitvec = "1"
 once_cell = "1"
+serde = "1"
+itertools = "0.10"
```
Comment (algesten) [inline str0m/Cargo.toml:13]: Do we really need this? I don't want more dependencies than we need. Does it solve a big problem?

File: str0m/src/lib.rs [hunk 3]
```diff
@@ -237,6 +245,12 @@ pub enum Event {
 
     /// A data channel has been closed.
     ChannelClose(ChannelId),
+
+    PeerStats(PeerStats),
+
+    MediaIngressStats(MediaIngressStats),
+
+    MediaEgressStats(MediaEgressStats),
```
Comment (algesten) [inline str0m/src/lib.rs:253]: These 3 events needs doc.

File: str0m/src/lib.rs [hunk 4]
```diff
@@ -989,15 +1018,32 @@ impl Rtc {
 
     fn do_handle_timeout(&mut self, now: Instant) {
         self.last_now = now;
+        info!("timeout");
```
Comment (algesten) [inline str0m/src/lib.rs:1021]: Remove

File: str0m/src/lib.rs [hunk 5]
```diff
@@ -989,15 +1018,32 @@ impl Rtc {
...
         self.ice.handle_timeout(now);
         self.sctp.handle_timeout(now);
         self.session.handle_timeout(now);
+        if self.stats.handles_timeout(now) {
```
Comment (algesten) [inline str0m/src/lib.rs:1025]: Let's rename this `wants_timeout`

File: str0m/src/media/mod.rs [hunk 6]
```diff
@@ -248,12 +248,13 @@ pub struct Media {
     /// These are created first time we observe the SSRC in an incoming RTP packet.
     /// Each source keeps track of packet loss, nack, reports etc. Receiving sources are
     /// cleaned up when we haven't received any data for the SSRC for a while.
-    sources_rx: Vec<ReceiverSource>,
+    // TODO: probably not ok to make them public
+    pub sources_rx: Vec<ReceiverSource>,
```
Comment (algesten) [inline str0m/src/media/mod.rs:252]: Very not ok. `Media` is public API and `ReceiverSource` isn't.

File: str0m/src/media/sender.rs [hunk 7]
```diff
@@ -10,9 +10,10 @@ use super::Source;
 pub struct SenderSource {
     ssrc: Ssrc,
     repairs: Option<Ssrc>,
-    rid: Option<Rid>,
+    pub rid: Option<Rid>,
```
Comment (algesten) [inline str0m/src/media/sender.rs:13]: There's accessor methods for rid.

File: str0m/src/media/receiver.rs [hunk 8]
```diff
@@ -16,6 +16,7 @@ pub struct ReceiverSource {
     sender_info: Option<SenderInfo>,
     sender_info_at: Option<Instant>,
     fir_seq_no: u8,
+    pub bytes_rx: u64,
```
Comment (algesten) [inline str0m/src/media/receiver.rs:19]: :(

File: str0m/src/stats.rs [hunk 9]
```diff
@@ -0,0 +1,231 @@
...
+
+use crate::{media::Source, Mid, Rtc};
+use rtp::Rid;
+
+pub struct Stats {
+    last_snapshot: StatsSnapshot,
+    events: VecDeque<StatEvent>,
+}
+
+type Bytes = u64;
```
Comment (algesten) [inline str0m/src/stats.rs:15]: This will be confusing for readers since a lot of people are familiar with the `bytes` crate. Let's just use u64.

**Review by algesten — CHANGES_REQUESTED** [review]
Looks quite good.

I think overall structure needs to change. Now we have this `StatsSnapshot` that requires access to very low level fields of each level – which in turn requires you to make a whole bunch of fields `pub`, that really shouldn't be pub.

It's especially bad when there's deep field dependencies two levels. Like Session level reaches all the way into a public field on a ReceiverSource.

Rather than manipulating fields directly I think you should pass the data via the in between structures and make proper functions for them.

I.e. `Media::update_rececived_bytes(&mut self, ssrc: Ssrc, len: u64)`. Which in turn calls `ReceiverSource::update_received_bytes(&mut self, len: u64)`.

Similar in the other direction when creating a report. Let each layer create the reports for that layer. `Media::get_stats_report(…)` and they in turn delegates to `ReceiverSource::get_stats_report(…)`.

File: str0m/Cargo.toml (hunk 2 above)
Comment (davibe) [reply str0m/Cargo.toml:13]: I used it for group_by() alone. Up to you.

Comment (algesten) [reply str0m/Cargo.toml:13]: Lets try without

Comment (davibe) [reply str0m/Cargo.toml:13]: done

Comment (davibe) [reply str0m/Cargo.toml]: done

File: rtp/Cargo.toml [hunk 10]
```diff
@@ -13,6 +13,7 @@ once_cell = "1"
 # instead of relying on multiple TLS libraries.
 openssl = { version = "0.10", features = ["vendored"] }
 openssl-sys = "^0.9"
+serde = "1"
```
Comment (algesten) [inline rtp/Cargo.toml:16]: Maybe a comment that we can move this dep if we move the id generation to a util crate?

File: str0m/src/media/mod.rs [hunk 11]
```diff
@@ -1193,6 +1198,31 @@ impl Media {
...
+            snapshot.tx += bytes_tx;
+            let key = (self.mid, s.rid());
+            if let Some(v) = snapshot.ingress.get_mut(&key) {
+                *v += bytes_tx;
+            } else {
+                snapshot.ingress.insert(key, bytes_tx);
+            }
+            // TODO: handle remote stats for this egress source
+        }
+    }
```
Comment (algesten) [inline str0m/src/media/mod.rs:1225]: Wouldn't it be cleaner to implement `visit_stats` on ReceiverSource and SenderSource here?

File: str0m/src/stats.rs [hunk 12]
```diff
@@ -0,0 +1,190 @@
...
+/// note: when simulcast is disabled, `rid` is `None`
+#[derive(Debug, Clone)]
+pub struct MediaEgressStats {
+    pub mid: Mid,
+    pub rid: Option<Rid>,
+
+    pub bitrate_tx: f32,
+    pub ts: Instant,
+    // TODO
+    pub remote: RemoteIngressStats,
```
Comment (algesten) [inline str0m/src/stats.rs:75]: Maybe comment this out?

File: str0m/src/stats.rs [hunk 13]
```diff
@@ -0,0 +1,190 @@
...
+/// note: when simulcast is disabled, `rid` is `None`
+#[derive(Debug, Clone)]
+pub struct MediaIngressStats {
+    pub mid: Mid,
+    pub rid: Option<Rid>,
+
+    pub bitrate_rx: f32,
+    pub ts: Instant,
+    // TODO
+    pub remote: RemoteEgressStats,
```
Comment (algesten) [inline str0m/src/stats.rs:94]: ditto

**Review by algesten — CHANGES_REQUESTED** [review]
Looks good! A couple of small questions.

File: str0m/src/stats.rs (hunk 12 above)
Comment (davibe) [reply str0m/src/stats.rs:75]: makes sense

File: str0m/src/media/mod.rs (hunk 11 above)
Comment (algesten) [reply str0m/src/media/mod.rs:1225]: Yeah, pass args in `visit_stats`. I like that.

**Review by algesten — APPROVED** [review]
Very nice!

**Pull request merged**